Can't mimick AWS SNS.publish

I'm trying to call my callback when I call `publish` on a Substituted AWS `SNS` instance. My code looks something like the sample below, but `sns.publish(Arg.any('object'),Arg.any())` doesn't seem to have the `mimick` method. I've played around with the argument matching and can't seem to get this to work. I don't know if this is an issue or a lack of understanding on my part.
```
import { SNS, AWSError } from 'aws-sdk';
import { Request } from 'aws-sdk/lib/request';
import { Substitute, Arg } from '@fluffy-spoon/substitute';

const sns = Substitute.for<SNS>();
function callback (data: SNS.PublishInput) {
    const snsResponse = Substitute.for<Request<SNS.PublishResponse, AWSError>>();
    console.log(`data=${data}, snsResponse=${snsResponse}`);
    return snsResponse;
}
sns.publish(Arg.any('object'),Arg.any()).mimicks(callback);
```

Ultimately what I want to do is to simulate different behavior/responses from my mocked `publish` method based on what was passed in the first parameter to the call to `publish` (which is of type `SNS.PublishInput`) I cannot also figure out how to check that a certain input was passed to `publish`. When I try to pass an object that adheres to the `SNS.PublishInput` interface to the   `received().publish()` check below

```
sns.received().publish(publishInput)
```
results in the following compiler error. I'm trying to verify that a particular input is passed to the `publish` call by the software under test.
```
error TS2769: No overload matches this call.
  Overload 1 of 2, '(callback?: ((err: AWSError, data: PublishResponse) => void) | undefined): void', gave the following error.
    Argument of type 'PublishInput' is not assignable to parameter of type '(err: AWSError, data: PublishResponse) => void'.
      Type 'PublishInput' provides no match for the signature '(err: AWSError, data: PublishResponse): void'.
  Overload 2 of 2, '(arg: AllArguments): void', gave the following error.
    Argument of type 'PublishInput' is not assignable to parameter of type 'AllArguments'.
      Type 'PublishInput' is missing the following properties from type 'AllArguments': description, matchingFunction, matches
```

--- a/src/Transformations.ts
+++ b/src/Transformations.ts
@@ -1,6 +1,6 @@
 import { AllArguments } from "./Arguments";
 
-type FunctionSubstituteWithOverrides<TFunc, Terminating = false> =
+type FunctionSubstituteWithOverloads<TFunc, Terminating = false> =
     TFunc extends {
         (...args: infer A1): infer R1;
         (...args: infer A2): infer R2;
@@ -76,7 +76,7 @@ type TerminatingObject<T> = {
     [P in keyof T]:
     T[P] extends (...args: infer F) => any ?
     F extends [] ? () => void :
-    FunctionSubstituteWithOverrides<T[P], true> :
+    FunctionSubstituteWithOverloads<T[P], true> :
     T[P];
 }
 
@@ -84,7 +84,7 @@ type ObjectSubstituteTransformation<T extends Object> = {
     [P in keyof T]:
     T[P] extends (...args: infer F) => infer R ?
     F extends [] ? NoArgumentFunctionSubstitute<R> :
-    FunctionSubstituteWithOverrides<T[P]> :
+    FunctionSubstituteWithOverloads<T[P]> :
     PropertySubstitute<T[P]>;
 }
 

--- a/src/Transformations.ts
+++ b/src/Transformations.ts
@@ -1,11 +1,46 @@
 import { AllArguments } from "./Arguments";
 
-export type NoArgumentFunctionSubstitute<TReturnType> = (() => (TReturnType & NoArgumentMockObjectMixin<TReturnType>))
+type FunctionSubstituteWithOverloads<TFunc> =
+    TFunc extends {
+        (...args: infer A1): infer R1;
+        (...args: infer A2): infer R2;
+        (...args: infer A3): infer R3;
+        (...args: infer A4): infer R4;
+        (...args: infer A5): infer R5;
+    } ?
+    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2> &
+    FunctionSubstitute<A3, R3> & FunctionSubstitute<A4, R4>
+    & FunctionSubstitute<A5, R5> : TFunc extends {
+        (...args: infer A1): infer R1;
+        (...args: infer A2): infer R2;
+        (...args: infer A3): infer R3;
+        (...args: infer A4): infer R4;
+    } ?
+    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2> &
+    FunctionSubstitute<A3, R3> & FunctionSubstitute<A4, R4> : TFunc extends {
+        (...args: infer A1): infer R1;
+        (...args: infer A2): infer R2;
+        (...args: infer A3): infer R3;
+    } ?
+    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2>
+    & FunctionSubstitute<A3, R3> : TFunc extends {
+        (...args: infer A1): infer R1;
+        (...args: infer A2): infer R2;
+    } ?
+    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2> : TFunc extends {
+        (...args: infer A1): infer R1;
+    } ?
+    FunctionSubstitute<A1, R1> : never;
+
+type Equals<A, B> = (<T>() => T extends A ? 1 : 2) extends (<T>() => T extends B ? 1 : 2) ? true : false;
 
 export type FunctionSubstitute<TArguments extends any[], TReturnType> =
+    Equals<TArguments, unknown[]> extends true ?
+    {} :
     ((...args: TArguments) => (TReturnType & MockObjectMixin<TArguments, TReturnType>)) &
     ((allArguments: AllArguments) => (TReturnType & MockObjectMixin<TArguments, TReturnType>))
 
+export type NoArgumentFunctionSubstitute<TReturnType> = (() => (TReturnType & NoArgumentMockObjectMixin<TReturnType>))
 export type PropertySubstitute<TReturnType> = (TReturnType & Partial<NoArgumentMockObjectMixin<TReturnType>>);
 
 type MockObjectPromise<TReturnType> = TReturnType extends Promise<infer U> ? {
@@ -41,8 +76,9 @@ type TerminatingObject<T> = {
 
 type ObjectSubstituteTransformation<T extends Object> = {
     [P in keyof T]:
-    T[P] extends (...args: infer F) => infer R ? FunctionSubstitute<F, R> :
-    T[P] extends () => infer R ? NoArgumentFunctionSubstitute<R> :
+    T[P] extends (...args: infer F) => infer R ?
+    F extends [] ? NoArgumentFunctionSubstitute<R> :
+    FunctionSubstituteWithOverloads<T[P]> :
     PropertySubstitute<T[P]>;
 }
 

--- a/src/Transformations.ts
+++ b/src/Transformations.ts
@@ -1,6 +1,6 @@
 import { AllArguments } from "./Arguments";
 
-type FunctionSubstituteWithOverloads<TFunc> =
+type FunctionSubstituteWithOverrides<TFunc, Terminating = false> =
     TFunc extends {
         (...args: infer A1): infer R1;
         (...args: infer A2): infer R2;
@@ -8,35 +8,38 @@ type FunctionSubstituteWithOverloads<TFunc> =
         (...args: infer A4): infer R4;
         (...args: infer A5): infer R5;
     } ?
-    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2> &
-    FunctionSubstitute<A3, R3> & FunctionSubstitute<A4, R4>
-    & FunctionSubstitute<A5, R5> : TFunc extends {
+    FunctionHandler<A1, R1, Terminating> & FunctionHandler<A2, R2, Terminating> &
+    FunctionHandler<A3, R3, Terminating> & FunctionHandler<A4, R4, Terminating>
+    & FunctionHandler<A5, R5, Terminating> : TFunc extends {
         (...args: infer A1): infer R1;
         (...args: infer A2): infer R2;
         (...args: infer A3): infer R3;
         (...args: infer A4): infer R4;
     } ?
-    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2> &
-    FunctionSubstitute<A3, R3> & FunctionSubstitute<A4, R4> : TFunc extends {
+    FunctionHandler<A1, R1, Terminating> & FunctionHandler<A2, R2, Terminating> &
+    FunctionHandler<A3, R3, Terminating> & FunctionHandler<A4, R4, Terminating> : TFunc extends {
         (...args: infer A1): infer R1;
         (...args: infer A2): infer R2;
         (...args: infer A3): infer R3;
     } ?
-    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2>
-    & FunctionSubstitute<A3, R3> : TFunc extends {
+    FunctionHandler<A1, R1, Terminating> & FunctionHandler<A2, R2, Terminating>
+    & FunctionHandler<A3, R3, Terminating> : TFunc extends {
         (...args: infer A1): infer R1;
         (...args: infer A2): infer R2;
     } ?
-    FunctionSubstitute<A1, R1> & FunctionSubstitute<A2, R2> : TFunc extends {
+    FunctionHandler<A1, R1, Terminating> & FunctionHandler<A2, R2, Terminating> : TFunc extends {
         (...args: infer A1): infer R1;
     } ?
-    FunctionSubstitute<A1, R1> : never;
+    FunctionHandler<A1, R1, Terminating> : never;
 
 type Equals<A, B> = (<T>() => T extends A ? 1 : 2) extends (<T>() => T extends B ? 1 : 2) ? true : false;
+type FunctionHandler<TArguments extends any[], TReturnType, Terminating> =
+    Equals<TArguments, unknown[]> extends true ?
+    {} : Terminating extends true ?
+    TerminatingFunction<TArguments> :
+    FunctionSubstitute<TArguments, TReturnType>
 
 export type FunctionSubstitute<TArguments extends any[], TReturnType> =
-    Equals<TArguments, unknown[]> extends true ?
-    {} :
     ((...args: TArguments) => (TReturnType & MockObjectMixin<TArguments, TReturnType>)) &
     ((allArguments: AllArguments) => (TReturnType & MockObjectMixin<TArguments, TReturnType>))
 
@@ -67,10 +70,13 @@ export type ObjectSubstitute<T extends Object, K extends Object = T> = ObjectSub
     mimick(instance: T): void;
 }
 
+type TerminatingFunction<TArguments extends any[]> = ((...args: TArguments) => void) & ((arg: AllArguments) => void)
+
 type TerminatingObject<T> = {
     [P in keyof T]:
-    T[P] extends (...args: infer F) => any ? ((...args: F) => void) & ((arg: AllArguments) => void) :
-    T[P] extends () => any ? () => void :
+    T[P] extends (...args: infer F) => any ?
+    F extends [] ? () => void :
+    FunctionSubstituteWithOverrides<T[P], true> :
     T[P];
 }
 
@@ -78,7 +84,7 @@ type ObjectSubstituteTransformation<T extends Object> = {
     [P in keyof T]:
     T[P] extends (...args: infer F) => infer R ?
     F extends [] ? NoArgumentFunctionSubstitute<R> :
-    FunctionSubstituteWithOverloads<T[P]> :
+    FunctionSubstituteWithOverrides<T[P]> :
     PropertySubstitute<T[P]>;
 }
 

